### ISSUE TITLE
use_build_context_synchronously is triggered when using context.mounted

### ISSUE THREAD

**Issue by luccasclezar** [dart-lang/linter#4007]
After upgrading to Flutter 3.7, using BuildContext.mounted in any way other than `if (context.mounted) {}` makes the linter complain about using context across async gaps.

@goderbauer told me to mention him.

# Reproduction steps
1. Create a new project with `flutter create`
2. Add this Widget to `main.dart`

```dart
class AsyncContextTestWidget extends StatelessWidget {
  const AsyncContextTestWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return ElevatedButton(
      onPressed: () async {
        await Future<void>.delayed(Duration.zero);

        // This line is marked as using context across async gaps
        if (!context.mounted) {
          return;
        }
      },
      child: const Text('Test'),
    );
  }
}
```

`!context.mounted` is just one example of this problem. This is reproducible when using ternary operators or initializing a variable to `context.mounted` too.

```dart
// This line triggers a warning too
final isMounted = context.mounted;

if (!isMounted) {
  return;
}
```

**Comment by andrewescutia** [comment]
If it helps here is an example failing test:

```
void methodWithBuildContextParameter4(BuildContext context) async { 
  await Future<void>.delayed(Duration());
  if (!context.mounted) {
    return;
  }
  await Navigator.of(context).pushNamed('routeName'); // OK
}
```

The added second `if` check below within `lib/src/rules/use_build_context_synchronously.dart` appears to correct it but I do not have an intimate knowledge of this repo.

```
        // if (mounted) { ... do ... }
        if (isMountedCheck(parent, positiveCheck: true)) {
          return;
        }

        // if (!context.mounted) { return; }
        if (isMountedCheck(parent)) {
          return;
        }
```

**Comment by LostInDarkMath** [comment]
This bug is pretty anoying. Hope you can fix this soon :)

**Comment by asashour** [comment]
For fixing this, should we search for specific statements, e.g. `if (!context.mounted) return;` or we should use FlowAnalysis, to check that `context` is not used before `.mount`, in the same function/method.

**Comment by pq** [comment]
It's unfortunate that we're approximating what flow analysis does more completely but I'm not aware of any way to leverage it in a lint implementation (though it has been discussed).

/fyi @stereotype441

**Comment by Levi-Lesches** [comment]
Just ran into this issue today, pretty unintuitive why 

```dart
if (!context.mounted) return;
```
was failing but 
```dart
if (context.mounted) doSomething(context)
```
wasn't

**Comment by luccasclezar** [comment]
@Levi-Lesches While this is not fixed, if you really want to return instead of placing more code inside an if, you can use:

```dart
if (context.mounted) {
} else {
  return;
}
```

It's not great, but sometimes it can be better than nesting ifs, plus when this is fixed the refactoring will be pretty quick.

**Comment by haslinger** [comment]
Cool tip, @luccasclezar! And to be able fo find these instances later, I modified it to
```dart
  if (context.mounted) {
    // FIXME
  } else {
    return;
  }
```

**Comment by JosephNK** [comment]
I hope it will be corrected. :)

**Comment by gnprice** [comment]
I'm continuing to see this issue as of a Dart SDK version from today (from the Flutter main/master channel):
```
$ dart --version
Dart SDK version: 3.1.0-75.0.dev (dev) (Thu May 4 03:59:43 2023 -0700) on "linux_x64"
```

In particular, if I take a test case [from the PR](https://github.com/dart-lang/linter/commit/85683badd16a48cf1d7bdea42ab7306f88465f72#diff-349f18cfb0b1630ee5c4fb89db47af62f0433c23910b4b56ddef13e6017c3ebd) #4312:
```dart
import 'package:flutter/widgets.dart';

void foo(BuildContext context) async {
  await f();
  if (!context.mounted) {
    return;
  }
  Navigator.of(context);
}

Future<void> f() async {}
```

and run `dart analyze`, I get the lint message:
```
$ dart analyze lib/tmp.dart
Analyzing tmp.dart...                  0.5s

   info • tmp.dart:5:8 • Don't use 'BuildContext's across async gaps. Try
          rewriting the code to not reference the 'BuildContext'. •
          use_build_context_synchronously

1 issue found.
```

Is that unexpected, or is there something I'm missing?

**Comment by pq** [comment]
@srawlins

**Comment by gnprice** [comment]
Relatedly: is there a straightforward way to look up what linter version is in the SDK? It'd be great to document that in the readme file too, if possible.

**Comment by srawlins** [comment]
All good questions!

I'd say we're aspiring for continuous deployment; each commit here should land ASAP in the Dart SDK. But we're definitely not there yet, commits here typically make it over to the Dart SDK in under a month.

> Relatedly: is there a straightforward way to look up what linter version is in the SDK? It'd be great to document that in the readme file too, if possible.

We use a homegrown vendoring system, so you'll see the commit SHA in this file: https://github.com/dart-lang/sdk/blob/main/DEPS#L147

**Comment by Wellcomezhangheng** [comment]
analyzer:
  errors:
    use_build_context_synchronously: ignore
Take away without thanking

**Comment by Wellcomezhangheng** [comment]
`
analyzer:
errors:
use_build_context_synchronously: ignore
`
Take away without thanking

**Comment by Wellcomezhangheng** [comment]
<img width="759" alt="image" src="https://github.com/dart-lang/linter/assets/16442105/c6394ebe-f2fc-4286-b82e-727c04f4f90c">
Take away without thanking
